src/auth/user-import-builder.ts

```diff
@@ -190,239 +285,192 @@ export class UserImportBuilder {
-            `hash algorithm ${this.options.hash.algorithm}.`,
+            `hash algorithm ${options.hash.algorithm}.`,
           );
         }
-        const blockSize = parseInt((this.options.hash.blockSize || '0').toString(), 10);
-        if (!validator.isNumber(this.options.hash.blockSize) || isNaN(blockSize)) {
+        const blockSize = getNumberField(options.hash, 'blockSize');
+        if (isNaN(blockSize)) {
           throw new FirebaseAuthError(
             AuthClientErrorCode.INVALID_HASH_BLOCK_SIZE,
             `A valid "hash.blockSize" number must be provided for ` +
-            `hash algorithm ${this.options.hash.algorithm}.`,
+            `hash algorithm ${options.hash.algorithm}.`,
           );
         }
-        const dkLen = parseInt((this.options.hash.derivedKeyLength || '0').toString(), 10);
-        if (!validator.isNumber(this.options.hash.derivedKeyLength) || isNaN(dkLen)) {
+        const dkLen = getNumberField(options.hash, 'derivedKeyLength');
+        if (isNaN(dkLen)) {
           throw new FirebaseAuthError(
             AuthClientErrorCode.INVALID_HASH_DERIVED_KEY_LENGTH,
             `A valid "hash.derivedKeyLength" number must be provided for ` +
-            `hash algorithm ${this.options.hash.algorithm}.`,
+            `hash algorithm ${options.hash.algorithm}.`,
           );
         }
-        this.validatedOptions = {
-          hashAlgorithm: this.options.hash.algorithm,
+        populatedOptions = {
+          hashAlgorithm: options.hash.algorithm,
           cpuMemCost,
           parallelization,
           blockSize,
           dkLen,
         };
         break;
+
       default:
         throw new FirebaseAuthError(
           AuthClientErrorCode.INVALID_HASH_ALGORITHM,
-          `Unsupported hash algorithm provider "${this.options.hash.algorithm}".`,
+          `Unsupported hash algorithm provider "${options.hash.algorithm}".`,
         );
     }
+    return populatedOptions;
   }
 
   /**
-   * Validates and populates the users list of the uploadAccount request.
+   * Validates and returns the users list of the uploadAccount request.
    * Whenever a user with an error is detected, the error is cached and will later be
    * merged into the user import result. This allows the processing of valid users without
    * failing early on the first error detected.
+   * @param {UserImportRecord[]} users The UserImportRecords to convert to UnploadAccountUser
+   *     objects.
+   * @param {ValidatorFunction=} userValidator The user validator function.
+   * @return {UploadAccountUser[]} The populated uploadAccount users.
    */
-  private validateAndPopulateUsers(): void {
+  private populateUsers(
+      users: UserImportRecord[], userValidator?: ValidatorFunction): UploadAccountUser[] {
     let index = 0;
-    this.users.forEach((user) => {
+    const populatedUsers: UploadAccountUser[] = [];
+    users.forEach((user) => {
```

Comment:
You could do `users.forEach((user, index) => {` here and avoid having to maintain you own `index` variable.

Reply:
done